libs/core/src/main/java/org/elasticsearch/common/unit/TimeValue.java

```diff
@@ -218,25 +247,68 @@ public String toString() {
             value = microsFrac();
             suffix = "micros";
         }
-        return formatDecimal(value) + suffix;
+        // Limit fraction pieces to a min of 0 and maximum of 10
+        return formatDecimal(value, Math.min(10, Math.max(0, fractionPieces))) + suffix;
     }
 
-    private static String formatDecimal(double value) {
+    private static String formatDecimal(double value, int fractionPieces) {
         String p = String.valueOf(value);
+        int totalLength = p.length();
         int ix = p.indexOf('.') + 1;
         int ex = p.indexOf('E');
-        char fraction = p.charAt(ix);
-        if (fraction == '0') {
+        // Location where the fractional values end
+        int fractionEnd = ex == -1 ? Math.min(ix + fractionPieces, totalLength) : ex;
+
+        // Determine the value of the fraction, so if it were .000 the
+        // actual long value is 0, in which case, it can be elided.
+        long fractionValue;
+        try {
+            fractionValue = Long.parseLong(p.substring(ix, fractionEnd));
+        } catch (NumberFormatException e) {
+            fractionValue = 0;
+        }
+
+        if (fractionValue == 0 || fractionPieces <= 0) {
+            // Either the part of the fraction we were asked to report is
+            // zero, or the user requested 0 fraction pieces, so return
+            // only the integral value
             if (ex != -1) {
                 return p.substring(0, ix - 1) + p.substring(ex);
             } else {
                 return p.substring(0, ix - 1);
             }
         } else {
+            // Build up an array of fraction characters, without going past
```

Comment:
I think this could be simplified to 
```
 } else {
            // truncate trailing zero's
            char[] fraction = p.substring(ix, fractionEnd).toCharArray();
            for (int i = fraction.length - 1; i > 0; i--) {
                if (fraction[i] == '0') {
                    fraction[i] = '\0';
                }
            }

            if (ex != -1) {
                return p.substring(0, ix) + new String(fraction) + p.substring(ex);
            } else {
                return p.substring(0, ix) + new String(fraction);
            }
        }
```
(I havent tested the above very well)

Reply:
EDIT ...nvmd that does work if there are zero's in the decimal but not trailing 🤦‍♂